### PR TITLE
Adds Wildcard Admission Policy

### PR DESCRIPTION
**Pull request by danehans**
Implements [Wildcard Admission Policy](https://github.com/openshift/enhancements/pull/252) enhancement proposal.

/assign @ironcladlou @knobunc 
/cc @Miciah @frobware 

/hold until https://github.com/openshift/api/pull/607 merges.

### CODE REVIEW

File: cmd/ingress-operator/start.go [hunk 1]
```diff
@@ -71,7 +71,7 @@ func NewStartCommand() *cobra.Command {
 }
 
 func start(opts *StartOptions) error {
-	metrics.DefaultBindAddress = ":60000"
+	metrics.DefaultBindAddress = ":60001"
```
Comment (ironcladlou) [inline cmd/ingress-operator/start.go]: What's this? Do we have some other unreported bug on master?

File: pkg/operator/controller/ingress/deployment.go [hunk 2]
```diff
@@ -436,16 +436,28 @@ func desiredRouterDeployment(ci *operatorv1.IngressController, ingressController
...
 		env = append(env, corev1.EnvVar{Name: "ROUTER_DISABLE_NAMESPACE_OWNERSHIP_CHECK", Value: "false"})
 	case operatorv1.InterNamespaceAllowedOwnershipCheck:
 		env = append(env, corev1.EnvVar{Name: "ROUTER_DISABLE_NAMESPACE_OWNERSHIP_CHECK", Value: "true"})
 	}
+	switch routeAdmission.WildcardPolicy {
+	case operatorv1.WildcardPolicyDisallowed:
+		env = append(env, corev1.EnvVar{Name: "ROUTER_ALLOW_WILDCARD_ROUTES", Value: "false"})
+	case operatorv1.WildcardPolicyAllowed:
+		env = append(env, corev1.EnvVar{Name: "ROUTER_ALLOW_WILDCARD_ROUTES", Value: "true"})
+	}
```
Comment (ironcladlou) [inline pkg/operator/controller/ingress/deployment.go]: Just FYI, during the namespace ownership implementation we noted that this approach will cause a rollout even though for default installs the setting hasn't changed, but we're okay with it (https://github.com/openshift/cluster-ingress-operator/pull/345#discussion_r366545076)

File: pkg/operator/controller/ingress/deployment.go [hunk 3]
```diff
@@ -436,16 +436,28 @@ func desiredRouterDeployment(ci *operatorv1.IngressController, ingressController
...
 	}
 	switch routeAdmission.NamespaceOwnership {
 	case operatorv1.StrictNamespaceOwnershipCheck:
 		env = append(env, corev1.EnvVar{Name: "ROUTER_DISABLE_NAMESPACE_OWNERSHIP_CHECK", Value: "false"})
 	case operatorv1.InterNamespaceAllowedOwnershipCheck:
 		env = append(env, corev1.EnvVar{Name: "ROUTER_DISABLE_NAMESPACE_OWNERSHIP_CHECK", Value: "true"})
 	}
+	switch routeAdmission.WildcardPolicy {
+	case operatorv1.WildcardPolicyDisallowed:
+		env = append(env, corev1.EnvVar{Name: "ROUTER_ALLOW_WILDCARD_ROUTES", Value: "false"})
```
Comment (ironcladlou) [inline pkg/operator/controller/ingress/deployment.go]: How about a constant?

Comment (ironcladlou) [reply pkg/operator/controller/ingress/deployment.go]: I realize I failed to introduce a constant for the other one...

File: cmd/ingress-operator/start.go (hunk 1 above)
Comment (danehans) [reply cmd/ingress-operator/start.go]: @ironcladlou I'm running the operator locally and port 6000 is already in use. I'll remove.

File: pkg/operator/controller/ingress/deployment.go (hunk 2 above)
Comment (danehans) [reply pkg/operator/controller/ingress/deployment.go]: Thanks for the heads-up.

File: pkg/operator/controller/ingress/deployment.go (hunk 3 above)
Comment (danehans) [reply pkg/operator/controller/ingress/deployment.go]: @ironcladlou I updated the PR to use a constant for "ROUTER_ALLOW_WILDCARD_ROUTES".

**Comment by ironcladlou** [comment]
Once rebased on https://github.com/openshift/api/pull/607 this lgtm

**Comment by danehans** [comment]
Also /hold until PR is rebased on https://github.com/openshift/cluster-ingress-operator/pull/384.

**Comment by danehans** [comment]
https://github.com/openshift/cluster-ingress-operator/pull/384 has merged.
/hold remove

**Comment by danehans** [comment]
https://github.com/openshift/cluster-ingress-operator/pull/384 has merged.
/hold cancel

File: pkg/operator/controller/ingress/deployment.go [hunk 4]
```diff
@@ -435,16 +439,28 @@ func desiredRouterDeployment(ci *operatorv1.IngressController, ingressController
...
+		}
 	}
 	switch routeAdmission.NamespaceOwnership {
 	case operatorv1.StrictNamespaceOwnershipCheck:
 		env = append(env, corev1.EnvVar{Name: "ROUTER_DISABLE_NAMESPACE_OWNERSHIP_CHECK", Value: "false"})
 	case operatorv1.InterNamespaceAllowedOwnershipCheck:
 		env = append(env, corev1.EnvVar{Name: "ROUTER_DISABLE_NAMESPACE_OWNERSHIP_CHECK", Value: "true"})
 	}
+	switch routeAdmission.WildcardPolicy {
+	case operatorv1.WildcardPolicyDisallowed:
```
Comment (Miciah) [inline pkg/operator/controller/ingress/deployment.go:459]: Should we have a `default:` case that either replaces or does a `fallthrough` to this case?

File: test/e2e/operator_test.go [hunk 5]
```diff
@@ -792,6 +792,60 @@ func TestRouteAdmissionPolicy(t *testing.T) {
...
+		t.Fatalf("failed to observe expected conditions: %v, status: %v", err, route3)
+	}
+
+	// Create a route with a wildcard policy of Subdomain.
+	route4Name := types.NamespacedName{Namespace: ns2.Name, Name: "route4"}
+	route4 := makeRoute(route4Name, "route4.test.example.com", "/bar")
+	route4.Spec.WildcardPolicy = routev1.WildcardPolicySubdomain
+
+	// The route should not be admitted because the ingresscontroller wildcard policy
+	// is None by default.
```
Comment (Miciah) [inline test/e2e/operator_test.go:819]: `None` → `WildcardsDisallowed`.

File: test/e2e/operator_test.go [hunk 6]
```diff
@@ -792,6 +792,60 @@ func TestRouteAdmissionPolicy(t *testing.T) {
...
+	if err := kclient.Update(context.TODO(), ic); err != nil {
+		t.Fatalf("failed to update ingresscontroller: %v", err)
+	}
+
+	// Recreate the route since the failed route will not automatically get
+	// readmitted and a route wildcard policy is immutable.
+	if err := kclient.Delete(context.TODO(), route4); err != nil {
+		t.Fatalf("failed to delete route: %v", err)
+	}
+	route4 = makeRoute(route4Name, "route4.test.example.com", "/bar")
```
Comment (Miciah) [inline test/e2e/operator_test.go:841]: You need to assign `route4.Spec.WildcardPolicy = routev1.WildcardPolicySubdomain` again.

File: pkg/operator/controller/ingress/deployment.go [hunk 7]
```diff
@@ -435,16 +439,28 @@ func desiredRouterDeployment(ci *operatorv1.IngressController, ingressController
...
+		WildcardPolicy:     operatorv1.WildcardPolicyDisallowed,
 	}
-	if admission := ci.Spec.RouteAdmission; admission != nil && len(admission.NamespaceOwnership) > 0 {
-		routeAdmission.NamespaceOwnership = ci.Spec.RouteAdmission.NamespaceOwnership
+	if admission := ci.Spec.RouteAdmission; admission != nil {
+		if len(admission.NamespaceOwnership) > 0 {
+			routeAdmission.NamespaceOwnership = ci.Spec.RouteAdmission.NamespaceOwnership
+		}
+		if len(admission.WildcardPolicy) > 0 {
+			routeAdmission.WildcardPolicy = ci.Spec.RouteAdmission.WildcardPolicy
```
[the remark this answers is not in the document]
Comment (danehans) [reply pkg/operator/controller/ingress/deployment.go]: done

File: pkg/operator/controller/ingress/deployment.go (hunk 4 above)
Comment (danehans) [reply pkg/operator/controller/ingress/deployment.go]: done

File: test/e2e/operator_test.go (hunk 5 above)
Comment (danehans) [reply test/e2e/operator_test.go]: done

File: test/e2e/operator_test.go (hunk 6 above)
Comment (danehans) [reply test/e2e/operator_test.go:841]: Good catch, thank you. Done.

File: test/e2e/operator_test.go [hunk 8]
```diff
@@ -792,6 +792,62 @@ func TestRouteAdmissionPolicy(t *testing.T) {
...
+	if err := kclient.Create(context.TODO(), route4); err != nil {
+		t.Fatalf("failed to create route: %v", err)
+	}
+	route4.Spec.WildcardPolicy = routev1.WildcardPolicySubdomain
```
Comment (Miciah) [inline test/e2e/operator_test.go:845]: This should be before the `kclient.Create`.

Comment (danehans) [reply test/e2e/operator_test.go]: Yeah, that would make sense. Too much multi-tasking on my end. I need to slow down and double-check my work. Thanks for catching this.

**Pull request merged**